lib/Dialect/FIRRTL/Transforms/ResolveTraces.cpp

```diff
@@ -75,49 +75,103 @@ struct ResolveTracesPass : public ResolveTracesBase<ResolveTracesPass> {
   /// runOnOperation.
   NLATable *nlaTable;
 
-  /// Internal implementation that updates an Annotation to add a "target" field
-  /// based on the current location of the annotation in the circuit.  The value
-  /// of the "target" will be a local target if the Annotation is local and a
-  /// non-local target if the Annotation is non-local.
-  bool updateTargetImpl(Annotation &anno, FModuleLike &module,
-                        FIRRTLBaseType type, StringRef name) {
-    if (!anno.isClass(traceAnnoClass))
-      return false;
+  /// Stores a pointer to an inner symbol table collection.
+  hw::InnerSymbolTableCollection *istc;
 
-    LLVM_DEBUG(llvm::dbgs() << "  - before: " << anno.getDict() << "\n");
+  /// Global symbol index used for substitutions, e.g., "{{42}}".  This value is
+  /// the _next_ index that will be used.
+  unsigned symbolIdx = 0;
 
-    SmallString<64> newTarget("~");
-    newTarget.append(module->getParentOfType<CircuitOp>().getName());
+  /// Map of symbol to symbol index.  This is used to reuse symbol
+  /// substitutions.
+  DenseMap<Attribute, unsigned> symbolMap;
+
+  /// Symbol substitutions for the JSON verbatim op.
+  SmallVector<Attribute> symbols;
+
+  /// Get a symbol index and update symbol datastructures.
+  unsigned getSymbolIndex(Attribute attr) {
+    auto iterator = symbolMap.find(attr);
+    if (iterator != symbolMap.end())
+      return iterator->getSecond();
+
+    auto idx = symbolIdx++;
+    symbolMap.insert({attr, idx});
+    symbols.push_back(attr);
+
+    return idx;
+  }
+
+  /// Convert an annotation path to a string with symbol substitutions.
+  void buildTarget(AnnoPathValue &path, SmallString<64> &newTarget) {
+
+    auto addSymbol = [&](Attribute attr) -> void {
+      newTarget.append("{{");
+      newTarget.append(std::to_string(getSymbolIndex(attr)));
```

Comment:
Please prefer using Twine -- this allocates a `std::string` just to throw it away, instead of emitting directly onto the vector (append Twine based on integer).

Reply:
I wanted to use `Twine`, but `SmallString::append` is only defined for `StringRef`. I figured going the `Twine(i).str()` was equivalent as that is also allocating an `std::string`. Any ideas?

Reply:
The `Twine(a).toVector(b)` appears to be the way.